### DISCUSSION TITLE
Handling large collections with `joined` load strategy

### DISCUSSION THREAD

**Discussion by efrias89** [mikro-orm/mikro-orm#5549] [Q&A]
Hi,

Using the now default `joined` strategy, the default `populateWhere: "all"`, and the following example entities:

```ts
@Entity()
export class Author {
    @PrimaryKey()
    id!: number

    @Property({ length: 255 })
    firstName: string

    @Property({ length: 255 })
    lastName: string

    @OneToMany(
        () => Book,
        book => book.author
    )
    books = new Collection<Book>(this)
}
```

```ts
@Entity()
export class Book {
    @PrimaryKey()
    id!: number

    @Property({ length: 255 })
    name: string

    @Property({ length: 255 })
    tag: string

    @ManyToOne({ entity: () => Author })
    author: Author
}
```

And after calling a `findAll` on the `Author` repository for example:

```ts
const authors = em.getRepository(Author).findAll({
    where: {
        books: {
            tag: "Fiction"
        }
    },
    populate: ["books"],
    limit: 10
})
```

It constructs the following query (formatted for some legibility):

```sql
select `a0`.*, `b1`.`id` as `b1__id`, `b1`.`name` as `b1__name`, `b1`.`tag` as `b1__tag`
from `author` as `a0` 
left join `books` as `b1` on `a0`.`id` = `b1`.`author_id` 
left join `books` as `b2` on `a0`.`id` = `b2`.`author_id` 
where `a0`.`id` in (
    select `a0`.`id` from (
        select `a0`.`id` 
        from `author` as `a0` 
        left join `books` as `b1` on `a0`.`id` = `b1`.`author_id` 
        left join `books` as `b2` on `a0`.`id` = `b2`.`author_id` 
        where `b2`.`tag` = 'Fiction' group by `a0`.`id` limit 10
    ) as `a0`
)
```

Which returns the expected result when processed by the ORM. But if I am not wrong, if there are 10 authors, and each author has 20 books for example, then that query is returning 4000 rows, which could be a lot bigger if the number of books by author increases and it could generate a meaningful performance hit.

Am I doing something wrong? or is there a reason that it couldn't be done another way?

**Comment by B4nan** [comment]
This is how SQL works, it's called a cartesian product. And because of this, the ORM employs a pagination mechanism - that's the subquery matching based on the PK you see in this example, since you can't just limit stuff that produces the cartesian product, you want to limit the root entities only. 

If you want to get around this, you can't use joins (or better say joins on to-many relations) - the alternative select-in strategy is here for exactly that case, it will use a query per type instead of joins (the joins will be still used for the where condition but not for the selects).

Do you have an actual issue or it's just you discovered how SQL joins work and you don't like that? :]

**Reply by efrias89** [reply]
The thing is that I was not expecting such a query. I was expecting something in the lines of:

```sql
select `a0`.*, `b1`.`id` as `b1__id`, `b1`.`name` as `b1__name`, `b1`.`tag` as `b1__tag`
from `author` as `a0` 
left join `books` as `b1` on `a0`.`id` = `b1`.`author_id` 
where `a0`.`id` in (
    select `a0`.`id` from (
        select `a0`.`id` 
        from `author` as `a0` 
        left join `books` as `b1` on `a0`.`id` = `b1`.`author_id` 
        where `b1`.`tag` = 'Fiction' group by `a0`.`id` limit 10
    ) as `a0`
)
```

Without a second `left join` to books. That's why I was unsure if there is room for optimizing the query it generated or there is some border case that I am not seeing that it must have a second `left join` to the same table in both queries.

**Reply by B4nan** [reply]
The second join is there because you use populateWhere: all, one join branch is for filtering the root entity, the other is for populating the relation - which needs all records, not just a subset based on the where query.

**Reply by efrias89** [reply]
Theoretically that makes sense because the first left join of the main query is used for populating and the second left join of the sub query is used for filtering. 

But I am trying to understand why the second `left join` (`b2`) in the main query is useful because it is not used for populating the relation nor for filtering the root entity, and why the first `left join` of the sub query should be done because it is also not used for populating nor for filtering.
